### DISCUSSION TITLE
Anytime Dynamic A* Path Planning Algorithm

### DISCUSSION THREAD

**Discussion by OrionSehn-personal** [networkx/networkx#6077] [Ideas]
## Adding Support for Anytime Dynamic Path Planning Algorithms

I am interested in implementing a python based anytime path planning algorithm as part of a university capstone project. Networkx seems the perfect opportunity to use others' work and meaningfully contribute to an open-source project. I wanted feedback on the idea to confirm that I am on the right track. 

### Details
The algorithm I intend to implement is detailed [here](http://www.cs.cmu.edu/~ggordon/likhachev-etal.anytime-dstar.pdf
)
#### My current thoughts on implementation:

I would add a dynamic path planning function in networkx\algorithms\shortest_paths\generic.py

Something like this:
def anytime_dynamic_shortest_path(G, source=None, target=None, path = None, weight=None, max_time = 60, method="astar"): 

This function would take in a partially optimized path, iterate on improving that path for a maximum time, then return the improved path. 

Write an implementation of anytime dynamic A* in the astar.py python file, or perhaps as its stand-alone anytime_dynamic_astar.py. 

I would add tests to verify this algorithm aligned with the code standards.

#### Please let me know if you have any thoughts or feedback on this implementation.

Orion

**Comment by dschult** [comment]
This looks quite interesting. We would certainly be interested in such a function.

I don't know the terms for this set of tools yet. You call it "anytime dynamic". I understand that to refer to: 
"anytime" - means that you cut off the search after a specific amount of time/work has been done
"dynamic" - means that the network could change. I'm not sure exactly what this means. Does each use of an anytime part of a search work with a static network and changes in the network occur between the partial anytime searches? I guess there are two features I am curious about: Can the network change while you are doing a part of the search? And can the suggested path that you provide to the function be no longer a valid path (e.g. because an edge is removed)?  Similarly is it just the weights of the edges that change or the existence of weights that change? 

Perhaps  the first things to decide are what to call the function and where to put it.  It looks like you are aiming for an astar version of the solution, so how about `anytime_dynamic_astar_path()` and put it into the module `astar.py`.  We can refactor if/when we get multiple anytime_dynamic algorithms.  This would also remove the `method` argument from the function because astar is the method.  Wouldn't you need a heuristic function input if you are using astar? And maybe a more descriptive input name than `path` would clarify that. Something like `initial_path`. The inputs source, target should be required and weight should default to `weight`, right?   So:  `anytime_dynamic_astar_path(G, source, target, heuristic=None, weight="weight", initial_path=None, max_time=60)`?

If the function goes into `astar.py` then the tests should go into `tests/test_astar.py`.  Thoughts?

**Reply by OrionSehn-personal** [reply]
Hello!

Thank you so much for your thoughtful and swift reply.

Your understanding of _anytime_ and _dynamic_ align with what I currently understand, though I am also still learning and exploring these algorithms myself. I believe that by _dynamic_, this implementation actually supports the network changes _while_ you are doing part of a search, and handles cases when paths no longer become valid. I figured that it would make sense to implement it thusly, and then if there was a use case where you needed to update the network in an automated fashion, then you could set the time operator to be as frequent as you wanted to update the path. 

I agree wholly with your suggestions. I also considered putting this function in astar.py, though I have a habit of trying to construct my code in such a way that minimizes the possibility of rework. However, you are totally correct, and there is not yet a need for this contingency, and reworking later is totally valid. Further going with the simpler implementation seems like a good way to get off to a good start. `initial_path` seems like a much better name for this starting path, you are correct I'll need a heuristic function to pass as well. 

I'll have to test a series of cases where consider cases where the graphs and paths change and conflict, and assure that they are handled. I think this will be somewhat painful.

**Comment by rossbar** [comment]
If "anytime" does indeed refer to computation time, that seems like a problem to me. Computation time is a nasty heuristic that would make something like this inherently unreproducible across systems (or even on the same system with different loads/resource allocation).

I'd be -1 on anything that incorporates computation time as a cutoff; however, a more concrete, reproducible metric might make sense, e.g. something like "number of edge swaps". In other words, an algorithm that tries to optimize an existing path subject to a constraint on a consistent, concrete metric derived from the network and/or path structure sounds more appealing to me.

I'm also not sure that "dynamic" as interpreted here is particularly relevant, as Python imposes some pretty stringent constraints on modifying certain data structures (e.g. dictionaries) while they're being iterated over. I suppose that may not be necessary - but I just thought I'd point it out to keep in mind while exploring the algorithm.

I'm not sure I understood correctly, so I apologize if I'm way out in left field!

**Comment by OrionSehn-personal** [comment]
Hello! 

I know this has been a long time!

I've finally gotten around to actually implementing this algorithm in NetworkX, and I'm thinking about how to merge this in, I was wondering about the best way to do it, and I wonder if I could get your opinions.

## Refresher:
Algorithm is based on the work done in [this paper](http://www.cs.cmu.edu/~ggordon/likhachev-etal.anytime-dstar.pdf)
- Anytime: The computed algorithm can be halted and a suboptimal solution produced. The optimality of the solution is dependent on a given parameter epsilon. This can be iteratively lowered to return more optimal solutions, where the work from previous iterations is reused. So it should be repeatable and hardware-agnostic. (ε = 1 is optimal and equivalent to a*, ε > 1 is suboptimal and higher ε is more suboptimal) @rossbar does this seem to address your concern on this front?

- Dynamic: The given graph can have changes made to it, this can happen while the algorithm is running. These changes must be made through another function that updates the dynamic tables core to the algorithm (OPEN/CLOSED/INCONS) 

## Example Execution

![image](https://github.com/networkx/networkx/assets/46799775/9e1ba42b-c302-4560-ae8a-5bc0ecc96b43)


> _Graphical Example of the following code_

```python 
    random.seed(1)
    G = nx.random_geometric_graph(100, 0.20, seed=896803)
    for (u, v, w) in G.edges(data=True): #Euclidean distance between nodes
        w['weight'] = np.sqrt((G.nodes[v]["pos"][0] - G.nodes[u]["pos"][0])**2 + (G.nodes[v]["pos"][1] - G.nodes[u]["pos"][1])**2)
    s_start, s_goal = 42, 25       
    def heursistic(u, v): #Euclidean distance between nodes
       return np.sqrt((G.nodes[v]["pos"][0] - G.nodes[u]["pos"][0])**2 + (G.nodes[v]["pos"][1] - G.nodes[u]["pos"][1])**2)
    
    # A* search for comparison
    start_time = time.time()
    path = nx.astar_path(G, s_start, s_goal, heursistic)
    print("A* time: ", time.time() - start_time)
    print("A* path: ", path)

    #create search object
    search = ADA_star(s_start, s_goal, G, heursistic)

    #compute first suboptimal path epsilon = 2
    start_time = time.time()
    search.compute_or_improve_path(epsilon=2)
    path = search.extract_path()
    print("ADA* epsilon = 2 time: ", time.time() - start_time)
    print("epsilon = 2 path: ", path)
    print("epsilon = 2 path_weight: ", nx.path_weight(G, search.extract_path(), "weight"))

    #compute second (better) suboptimal path
    search.compute_or_improve_path(epsilon=1.2)
    path = search.extract_path()
    print("epsilon = 1.2 path: ", path)
    print("epsilon = 1.2 path_weight: ", nx.path_weight(G, path, "weight"))

    #compute third (best) suboptimal path
    search.compute_or_improve_path(epsilon=1)
    path = search.extract_path()
    print("epsilon = 1 path: ", path)
    print("epsilon = 1 path_weight: ", nx.path_weight(G, path, "weight"))

    #change graph edge weight
    print("changing graph weight for edge (49, 97)")
    search.update_graph([[49, 97, 0]]) #add edge between 77 and 15 with weight 0
    search.compute_or_improve_path(epsilon=1)
    path = search.extract_path()
    print("changed epsilon = 1 path: ", path)
    print("changed epsilon = 1 path_weight: ", nx.path_weight(G, path, "weight"))
```

### Output
``` 
A* time:  0.00012803077697753906
A* path:  [42, 32, 19, 72, 49, 29, 31, 94, 35, 25]
ADA* epsilon = 2 time:  0.0009984970092773438
epsilon = 2 path:  [42, 32, 24, 40, 59, 4, 66, 27, 35, 25]
epsilon = 2 path_weight:  1.4679609830956495
epsilon = 1.2 path:  [42, 32, 24, 12, 59, 4, 1, 27, 35, 25]
epsilon = 1.2 path_weight:  1.3335657361796027
epsilon = 1 path:  [42, 32, 19, 72, 49, 29, 31, 94, 35, 25]
epsilon = 1 path_weight:  1.29129785933092
changing graph weight for edge (49, 97)
changed epsilon = 1 path:  [42, 32, 19, 72, 49, 97, 11, 31, 94, 35, 25]
changed epsilon = 1 path_weight:  1.1428811257616596
```
Note: The time to run suboptimal ADA* is quite a bit shorter than A*, and work can be recycled to improve paths. This time is much more noticeable in larger graphs. 

The idea would be that users could use a for loop to iteratively lower the epsilon and get results as often as they need to, and update the graph during these iterations.

## Implementation:
It's available here: [on my fork of the project](https://github.com/OrionSehn-personal/networkx/blob/ada_star/networkx/algorithms/shortest_paths/ada_star.py)
Some more comments need to be made for clarity and testing still has to be done, but you can look at it here for now. I would probably also appreciate advice on where exactly to merge this in. 

## Questions
- Does this still seem useful? and is this search object way of implementing the algorithm useful? I figured people could put this in a for loop which just decreases epsilon at their preferred decent rate, and fetch better paths when would be best for their use case. 
- How can I parameterize the changes to the graph better? This algorithm can handle additions or removals to the graph as well, but I must know about what nodes/edges changed, is there some graph diff? 
- Do you have any other feedback on this implementation before I go writing unit tests for it?
- Where do you think this might fit best? just throw it in the a_star.py file? 
- Is there anything else I can do to help this be useful? 



I eagerly await your responses 

Edit: submitted wrong picture.

**Reply by dschult** [reply]
Thanks for this!  

I don't know enough about this kind of problem to guess whether this is useful or not -- if others need this or have perspectives, please jump in. :)

The main things I'm wrestling with for this algorithm is how to handle the graph changes. If you have to know what nodes or edges have changed, then you'll need an input argument to that effect. It is a very difficult problem to even check if two graphs are isomorphic, let alone what the differences are. But I guess you don't need isomorphism checking --- just which edges or nodes are different. If that is the case you could take the set-difference of the nodes and edges (be careful that the direction of the edges is handled correctly). Or you could require the user to track the changes and input a list of nodes and edges that have been added and another for removed. And I suppose you could subclass the graph classes and have the graph structure track the changes since the last call -- that would cause you to make whole new graph data structures for this algorithm.  It all sounds fairly messy. Perhaps it would be best to make the user track the changes and submit them as inputs.

Some changes I would suggest that we would ask for, but don't affect anything "real" in the algorithm:
- we don't use inline type annotations (i.e. `--> None`) so those should be removed.
- the name `s` for nodes should be morphed to match the library. Usually `n` or `node` or, if there are two of them, `u` and `v` or `node` and `nbr` -- or even `n` and `nbr`.  Similarly `s_start` and `s_goal` are more consistent as `source` and `target`.
- the "private methods" that the user is not intended to use should start with a single underscore (python private functions (dunders) start and end with double underscores, but library private vars/function/methods aren't supposed to use those in case python wants to use that name -- so we just use `_do_something` as  the method name for private names. I find it easier to read than the double-underscore anyway.
- for the examples in the docstring, the comments can/should usually be written as text rather than code comments. The code is being embedded into text already in the docs.  And results like floating point values or timing results can't be included in the doc_string examples because the user won't  get those exact results. We test the doc_strings to ensure they produce correct results, and the test is a text comparison of the output with what appears in the doc_string. 

I think putting the code into `astar.py` with tests added to `tests/test_astar.py` would be a good way to approach this.

**Reply by OrionSehn** [reply]
Hello!

Thanks for your prompt response! 

I've implemented almost all the advice you've given.

The one exception is that I've implemented this as ada_star.py and placed tests under tests/test_ada_star.py. I thought this might be a significant enough departure from astar to warrant a different file, also it simplifies merge conflicts and upstream changes for me for the time being. I can merge this into astar.py if you still think that'd be better. 

I don't totally have a better idea for how to allow the user to input changes at this time, except to allow the user to input 3 new list parameters to the update graph method (new, edited, deleted). I'll likely implement this soon, but just updating edges lets the user set edges to float infinity right now so they can still remove some paths from the calculation. 

The Algorithm does not work with directed graphs right now, but I think this very well could be fixed. 

[I also noticed a typo in the weighted.py docstring](https://github.com/OrionSehn-personal/networkx/commit/15dffa2fd4c1625a0b6fd1c2f94b793cc41374ce) 


Let me know if moving into astar.py is a must, and if you had any other ways I could help out.

**Reply by dschult** [reply]
A separate module should be fine if it makes a PR easier.

**Reply by OrionSehn** [reply]
I've created a pull request for this:  [https://github.com/networkx/networkx/pull/6896](https://github.com/networkx/networkx/pull/6896)

Thank you again

**Comment by OrionSehn-personal** [comment]
Also what are you supposed to do if the Black linter gives you a suggestion which is obviously worse for readability than what you currently have? 

![image](https://github.com/networkx/networkx/assets/46799775/45434bb1-ccb1-4ed8-8f2b-7764535f988a)

Do you make an exception, or accept the worse but consistent version?
